Ext2 write + growing inode support

Supercip971 opened the pull request.

Supercip971 committed:
ext2: fixed large read issue + added Inode cache
- now we cache inode blocks and we cache the inode itself, avoiding some reads

Supercip971 committed:
ext2: added inode writing + growing

Supercip971 committed:
brutal: added flags to io_file_open
- now you can write / read a file

Supercip971 committed:
ext2: fixed inode growing

Supercip971 committed:
ext2: tyding up

sleepy-monax commented on sources/libs/embed/posix/file.c:
```diff
@@ -9,11 +9,27 @@
+        case FILE_OPEN_READ_WRITE:
+            return O_RDWR;
+        default:
+            panic$("unkown embed flag: {}", flags);
```
Unkown open flag: {}

sleepy-monax commented on sources/libs/fs/ext2/ext2.c:
```diff
@@ -2,12 +2,47 @@
 #include <fs/ext2/ext2.h>
 
-static void *ext2_acquire_blocks(Ext2Fs *self, size_t off, size_t count, bool write)
+static Ext2CachedInode *get_cached_inode(Ext2Fs *self, Ext2FsInodeId inode)
+{
+    vec_foreach(cached, &self->inode_cache)
+    {
+        if (cached->inode.id == inode)
+        {
+            return cached;
+        }
+    }
+    return nullptr;
+}
+
+static Ext2CachedInode *add_cached_inode(Ext2Fs *self, Ext2FsInode inode)
+{
+    Ext2CachedInode cached = {
+        .inode = inode,
+    };
+
+    vec_init(&cached.block, self->alloc);
+
+    vec_push(&self->inode_cache, cached);
+    return vec_end(&self->inode_cache);
+}
+
+static Ext2CachedInode *get_cached_inode_or_create(Ext2Fs *self, Ext2FsInode inode)
+{
+    Ext2CachedInode *cached = get_cached_inode(self, inode.id);
+    if (cached == nullptr)
+    {
+        return add_cached_inode(self, inode);
+    }
+
+    return cached;
+}
```
The inode cache should live it's own translation unit

sleepy-monax commented on sources/libs/fs/ext2/ext2.c:
```diff
@@ -43,120 +100,417 @@ static size_t get_entry_in_indirect_block(size_t entry_per_block, int depth)
     return len;
 }
 
+FsResult ext2_inode_rw(Ext2Fs *self, Ext2FsInode *inode, Ext2FsInodeId id, bool write)
```
I don't like function behaving completely differently based on a boolean, most of the time this mean the fonctions need to be splitted

Supercip971 replied:
It's to avoid repetitive code, and this function is splitted in 2: ext2_write_inode & ext2_inode

sleepy-monax replied:
Well then just isolate the common code in another function

Supercip971 replied:
I don't see the point of doing it if in the end we only use it in 2 functions.  (it should be static tho)

sleepy-monax commented on sources/libs/fs/ext2/ext2.c:
```diff
@@ -43,120 +100,417 @@ static size_t get_entry_in_indirect_block(size_t entry_per_block, int depth)
+    return FS_SUCCESS;
+}
+
+static FsResult ext2_alloc_block(Ext2Fs *self, uint32_t preferred_group_id, uint32_t *result)
```
block allocation should live in it's own TU

sleepy-monax commented on sources/libs/fs/fs.h:
```diff
@@ -8,4 +8,5 @@ typedef enum
     FS_INVALID_PATH,
     FS_INVALID_FS,
+    FS_NO_FREE_BLOCK
 } FsResult;
```
The Fs subsystem should use IoResult, making interop between our IO and FS subsystem easier.

Supercip971 committed:
file: split file_open in 4 functions
- file_init for custom flags
- file_view for readonly
- file_edit for read-write
- file_create for write only
- Change the name of flags
- Changed flags value
- added some doxygen

Supercip971 committed:
fs: use Result(T) instead of value for returning error
+ some fix in the ext2 driver:
- check write
- more error checking instead of just doing like nothing happened

Supercip971 committed:
meta: update roadmap

Supercip971 committed:
fs: split ext2.c in multiple files
- now we have ext2.c cache.c and block.c
not shure what should go to block.c because there is a lot of block-related function that are also related to inode or groups.
- used clang-format
- removed MAYBE_UNUSED of fstools main.c
- fixed null dereference when writing with a non existant inode
